test/e2e/vmagent_test.go

```diff
@@ -0,0 +1,307 @@
+		require.Nil(t, err, "Failed to create ExternalNode")
+		startAntreaAgent(t, data, vm)
+	}
+	return vmList, nil
+}
+
+// teardownVMAgentTest deletes ExternalNode, verifies ExternalEntity is deleted
+// and verifies uplink configuration is restored.
+func teardownVMAgentTest(t *testing.T, data *TestData, vmList []vmInfo) {
+	verifyUpLinkAfterCleanup := func(vm vmInfo) {
+		if err := wait.PollImmediate(30*time.Second, 1*time.Minute, func() (done bool, err error) {
+			var tempVm vmInfo
+			if vm.osType == "Linux" {
+				tempVm = getVMInfo(t, data, vm.nodeName)
+			} else {
+				tempVm = getWindowsVMInfo(t, data, vm.nodeName)
+			}
+			if vm.ifName != tempVm.ifName {
+				t.Errorf("Failed to verify uplink interface, Expected %s, Got %s", vm.ifName, tempVm.ifName)
+				return false, nil
+			}
+			if vm.ip != tempVm.ip {
+				t.Errorf("Failed to verify uplink IP, Expected %s, Got %s", vm.ip, tempVm.ip)
+				return false, nil
+			}
+			return true, nil
+		}); err != nil {
+			assert.Fail(t, "Failed to verify uplink configuration after cleanup")
+			return
+		}
+	}
+	t.Logf("TestVMAgent teardown")
+	for _, vm := range vmList {
+		hash := sha1.New()
+		io.WriteString(hash, vm.ifName)
+		hashedIfName := hex.EncodeToString(hash.Sum(nil))
+		// Generate ExternalEntity name from ExternalNode
+		eeName := vm.nodeName + "-" + hashedIfName[:interfaceNameLength]
+		err := data.crdClient.CrdV1alpha1().ExternalNodes(nameSpace).Delete(context.TODO(), vm.nodeName, metav1.DeleteOptions{})
+		require.Nil(t, err, "Failed to delete ExternalNode")
+		verifyExternalEntityExistence(t, data, eeName, vm.nodeName, false)
+		verifyUpLinkAfterCleanup(vm)
+	}
+}
+
+func verifyExternalEntityExistence(t *testing.T, data *TestData, eeName string, vmNodeName string, expectExists bool) {
+	if err := wait.PollImmediate(10*time.Second, 1*time.Minute, func() (done bool, err error) {
+		t.Logf("Verifying ExternalEntity %s, expectExists %t", eeName, expectExists)
+		_, err = data.crdClient.CrdV1alpha2().ExternalEntities(nameSpace).Get(context.TODO(), eeName, metav1.GetOptions{})
+		if err != nil && !errors.IsNotFound(err) {
+			t.Errorf("Failed to get ExternalEntity %s by ExternalNode %s: %v", eeName, vmNodeName, err)
+			return false, err
+		}
+
+		if expectExists {
+			if err != nil {
+				return false, err
+			}
+			return true, nil
+		}
+		// ExternalEntity is expected to be deleted, check for error
+		if err != nil {
+			return true, nil
+		}
+		return false, nil
+
+	}); err != nil {
+		op := "created"
+		if !expectExists {
+			op = "deleted"
+		}
+		assert.NoError(t, err, "Failed to verify ExternalEntity %s %s by ExternalNode %s", eeName, op, vmNodeName)
+	}
+}
+
+func testExternalNode(t *testing.T, data *TestData, vmList []vmInfo) {
+	verifyExternalNodeRealization := func(vm vmInfo) {
+		if err := wait.PollImmediate(30*time.Second, 1*time.Minute, func() (done bool, err error) {
+			t.Logf("Verify host interface configuration, VmInfo %+v", vm)
```

Comment:
ditto

Reply:
Here it's dumping vmInfo object, because we are verifying Interfaces are present in OVS
The logs will contain the interface name that is to be present, which would help in troubleshooting.